pkg/engine/internal/planner/logical/planner.go

```
@@ -214,132 +216,176 @@ func buildPlanForLogQuery(
-				groupBy = append(groupBy, *NewColumnRef(group, types.ColumnTypeAmbiguous))
-			}
+func walkVectorAggregation(e *syntax.VectorAggregationExpr, params logql.Params) (Value, error) {
+	// `without()` grouping is not supported.
+	if e.Grouping != nil && e.Grouping.Without {
+		return nil, errUnimplemented
+	}
 
-			return true
-		default:
-			err = errUnimplemented
-			return false // do not traverse children
-		}
-	})
+	left, err := walk(e.Left, params)
 	if err != nil {
 		return nil, err
 	}
 
-	if rangeAggType == types.RangeAggregationTypeInvalid || vecAggType == types.VectorAggregationTypeInvalid {
+	vecAggType := convertVectorAggregationType(e.Operation)
+	if vecAggType == types.VectorAggregationTypeInvalid {
 		return nil, errUnimplemented
 	}
 
-	logSelectorExpr, err := e.Selector()
+	groupBy := make([]ColumnRef, 0, len(e.Grouping.Groups))
+	for _, group := range e.Grouping.Groups {
+		groupBy = append(groupBy, *NewColumnRef(group, types.ColumnTypeAmbiguous))
+	}
+
+	return &VectorAggregation{
+		Table:     left,
+		GroupBy:   groupBy,
+		Operation: vecAggType,
+	}, nil
+}
+
+func hasNonMathExpressionChild(n Value) bool {
+	if _, ok := n.(*VectorAggregation); ok {
+		return true
+	}
+
+	if _, ok := n.(*RangeAggregation); ok {
+		return true
+	}
+
+	if b, ok := n.(*BinOp); ok {
+		return hasNonMathExpressionChild(b.Left) || hasNonMathExpressionChild(b.Right)
+	}
+
+	return false
+}
+
+func walkBinOp(e *syntax.BinOpExpr, params logql.Params) (Value, error) {
+	left, err := walk(e.SampleExpr, params)
 	if err != nil {
 		return nil, err
 	}
-
-	builder, err := buildPlanForLogQuery(logSelectorExpr, params, true, rangeInterval)
+	right, err := walk(e.RHS, params)
 	if err != nil {
 		return nil, err
 	}
 
-	// Add unwrap instruction if present
-	if unwrapIdentifier != "" {
-		builder = builder.Cast(unwrapIdentifier, unwrapOperation)
+	op := convertBinaryArithmeticOp(e.Op)
+	if op == types.BinaryOpInvalid {
+		return nil, errUnimplemented
```

An invalid op type should return an error.